### ISSUE TITLE
Baseline results for lightly-modified (Tensorflow and PyTorch) TransE examples?

### ISSUE THREAD

**Issue by drcdr** [thunlp/OpenKE#179]
**SUMMARY**
a) I'm getting started, trying to test example_train_transe.py for both OpenKE TF and PyTorch
b) Tensorflow seems OK; but maybe there is a issue in the PyTorch version??
c) There may be similarities to https://github.com/thunlp/OpenKE/issues/133, but that was closed without comment.

**DETAILS**
I would like to know what results others are getting for some baseline examples.  PyTorch results in particular look wrong; the loss scores are bouncing around; many 'bad count's; very poor results.  Both of these examples use FB15K and SGD.  I'm using Python 3.6, and:
* For Tensorflow:    tf-nightly-gpu            1.13.0.dev20190214
* For PyTorch:        pytorch-nightly           1.1.0.dev20190515

I made some minor print formatting changes to Config.py, and added `con.test_flag = True`.  For the PyTorch test, I tried both 10 and 250 for save_steps, valid_steps, & early_stopping_patience.

My results:
**OpenKE-Tensorflow, example_train_transe.py:**  
```
Epoch:   0, loss:    0.1052090, time:   0.7
Epoch:   1, loss:    0.1007838, time:   0.5
...
Epoch: 998, loss:    0.0596958, time:   0.6
Epoch: 999, loss:    0.0687566, time:   0.5
no type constraint results:
metric:                  MRR             MR              hit@10          hit@3           hit@1
l(raw):                  0.736897        1964.592896     0.737265        0.737265        0.736520
r(raw):                  0.779142        1651.600220     0.779147        0.779147        0.779130
averaged(raw):           0.758020        1808.096558     0.758206        0.758206        0.757825

l(filter):               0.736897        1915.440552     0.737265        0.737265        0.736520
r(filter):               0.779142        1620.453003     0.779147        0.779147        0.779130
averaged(filter):        0.758020        1767.946777     0.758206        0.758206        0.757825
type constraint results:
metric:                  MRR             MR              hit@10          hit@3           hit@1
l(raw):                  0.741827        217.904861      0.746170        0.741040        0.738129
r(raw):                  0.783773        178.922348      0.788576        0.782584        0.780231
averaged(raw):           0.762800        198.413605      0.767373        0.761812        0.759180

l(filter):               0.757100        168.794678      0.770666        0.757292        0.750030
r(filter):               0.787800        147.711533      0.797633        0.787696        0.782448
averaged(filter):        0.772450        158.253113      0.784150        0.772494        0.766239
triple classification accuracy is 0.500076
The total of test triples is 59071.
The total of valid triples is 50000.
[14887 10571 11515  7820  1704]
The total of test triples is 59071.
The total of valid triples is 50000.
[1506 3468 2528 2538 9754]
The total of test triples is 59071.
The total of valid triples is 50000.
[  9  61 167 128 203]
The total of test triples is 59071.
The total of valid triples is 50000.
triple (151,152,9) is wrong
The total of test triples is 59071.
The total of valid triples is 50000.
triple (151,152,8) is wrong
```

**OpenKE-PyTorch, example_train_transe.py:** 
_[I'm getting similar behavior with OpenKE/PyTorch/examples/train_transE.py.]_

```
Epoch:   0, loss: 6434.0659180, time:   1.6
Epoch:   1, loss:   71.7901306, time:   1.5
Epoch:   2, loss:   61.7268677, time:   1.5
Epoch:   3, loss:   58.3037415, time:   1.5
Epoch:   4, loss:   57.2297058, time:   1.5
Epoch:   5, loss:   64.1686096, time:   1.5
...
Epoch: 990, loss:    1.8808594, time:   1.6
Epoch: 991, loss:    0.9441528, time:   1.6
Epoch: 992, loss:    0.6458130, time:   1.6
Epoch: 993, loss:    0.0000000, time:   1.6
Epoch: 994, loss:    0.0000000, time:   1.6
Epoch: 995, loss:    0.0000000, time:   1.7
Epoch: 996, loss:    0.0000000, time:   1.6
Epoch: 997, loss:    2.3342896, time:   1.5
Epoch: 998, loss:    0.0000000, time:   1.6
Epoch: 999, loss:    0.8283691, time:   1.6
Epoch 999 has finished, saving...
Epoch 999 has finished, validating 50000...
Hit@10 of valid set is 0.001420 | bad count is 1
Best epoch is 499 | hit@10 of valid set is 0.001430
Store checkpoint of best result at epoch 499...
Finish storing
Testing...
Initializing test model...
Finish initializing
The total of test triple is 59071
no type constraint results:
metric:                  MRR             MR              hit@10          hit@3           hit@1
l(raw):                  0.001022        4706.615234     0.000948        0.000322        0.000119
r(raw):                  0.001309        4542.170898     0.001642        0.000525        0.000135
averaged(raw):           0.001166        4624.393066     0.001295        0.000423        0.000127

l(filter):               0.001037        4618.001465     0.000982        0.000339        0.000119
r(filter):               0.001317        4485.715820     0.001642        0.000525        0.000135
averaged(filter):        0.001177        4551.858398     0.001312        0.000432        0.000127
type constraint results:
metric:                  MRR             MR              hit@10          hit@3           hit@1
l(raw):                  0.062749        456.430878      0.121024        0.063110        0.025004
r(raw):                  0.100627        383.157898      0.179191        0.105077        0.053732
averaged(raw):           0.081688        419.794373      0.150108        0.084094        0.039368

l(filter):               0.104877        367.470734      0.186640        0.107853        0.058912
r(filter):               0.125503        326.513641      0.214488        0.130216        0.075841
averaged(filter):        0.115190        346.992188      0.200564        0.119035        0.067377
triple classification accuracy is 0.516247

```

**QUESTIONS**
1) Are other getting different results?
2) Should MRR, hit@X be the same for raw and filtered, for no-type-constraint?
3) Are others seeing that PyTorch takes 3x as long as Tensorflow?
4) Is the PyTorch 'loss-bouncing' normal?

**Comment by ShulinCao** [comment]
Can you show your example_train_transe.py?

**Comment by drcdr** [comment]
Here is the PyTorch version:
```
import config
from  models import *
import json
import os 
os.environ['CUDA_VISIBLE_DEVICES']='0'  # was 6
con = config.Config()
con.set_in_path("./benchmarks/FB15K/")
con.set_work_threads(1) # was 8
con.set_train_times(1000)
con.set_nbatches(100)	
con.set_alpha(0.001)
con.set_bern(0)
con.set_dimension(100)
con.set_margin(1.0)
con.set_ent_neg_rate(1)
con.set_rel_neg_rate(0)
con.set_opt_method("SGD")
con.test_flag = True  # added

N = 250 # tried both N=10 and N=250
con.set_save_steps(N)
con.set_valid_steps(N)
con.set_early_stopping_patience(N)
con.set_checkpoint_dir("./checkpoint")
con.set_result_dir("./result")
con.set_test_link(True)
con.set_test_triple(True)
con.init()
con.set_train_model(TransE)
con.train()

con.test()  # added

```
Here is the Tensorflow version:
```
import config
import models
import tensorflow as tf
import numpy as np
import os
os.environ['CUDA_VISIBLE_DEVICES']='0'  # was 7
#Input training files from benchmarks/FB15K/ folder.
con = config.Config()
#True: Input test files from the same folder.
con.set_in_path("./benchmarks/FB15K/")
con.set_test_link_prediction(True)
con.set_test_triple_classification(True)
con.set_work_threads(1) # was 8
con.set_train_times(1000)
con.set_nbatches(100)
con.set_alpha(0.001)
con.set_margin(1.0)
con.set_bern(0)
con.set_dimension(100)
con.set_ent_neg_rate(1)
con.set_rel_neg_rate(0)
con.set_opt_method("SGD")
con.test_flag = True  # added

#Models will be exported via tf.Saver() automatically.
con.set_export_files("./res/model.vec.tf", 0)
#Model parameters will be exported to json files automatically.
con.set_out_files("./res/embedding.vec.json")
#Initialize experimental settings.
con.init()
#Set the knowledge embedding model
con.set_model(models.TransE)
#Train the model.
con.run()
#To test models after training needs "set_test_flag(True)".
con.test()
con.predict_head_entity(152, 9, 5)
con.predict_tail_entity(151, 9, 5)
con.predict_relation(151, 152, 5)
con.predict_triple(151, 152, 9)
con.predict_triple(151, 152, 8)
```

**Comment by ShulinCao** [comment]
![image](https://user-images.githubusercontent.com/19309481/60398485-40b6ec00-9b8b-11e9-8561-8f9dc85d97a9.png)
I tried your script and It's ok.

**Comment by drcdr** [comment]
Thanks for running this @ShulinCao.  We got different results for the OpenKE/PyTorch version, so I'll try and figure out why.  (Do you have any suggestions for how to do this?)  Also, what version of PyTorch are you using?  And, I assume you ran this off of the head of branch OpenKE-PyTorch?

But it looks like you are getting substantially lower performance than the OpenKE/Tensorflow version of the same script, see above.  Do you know why that may be?

**Comment by ShulinCao** [comment]
1. The pytorch version is 0.4.1. I just cloned the OpenKE-PyTorch branch and run the code. 
2. I don't quite understand "substantially lower performance". It seems the pytorch version is slightly better... And I believe the results of raw setting and filter setting should be different. There must be something wrong with your codes.

**Comment by drcdr** [comment]
My issue is that I was effectively building the C++ base software in 32-bit mode instead of 64-bit mode.  That's fixed.  I now have both the TF and PyT versions running on both Linux and Win10 (MSVC).  

Here is a summary of the results I'm now getting, for :  FB15K, TransE, Averaged/Filtered, No Type Constraint


N | OS | FW | Opt | Loss,0 | Loss,950 | MRR | MR | hit@10 | hit@3 | hit@1
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
1 | Linux | TF | SGD | 108.33 | 7.599 | 0.0924 | 804.4385 | 0.2440 | 0.1291 | 0.0077
2 | Linux | TF | Adam | 54.003 | 0.2015 | 0.3608 | 92.3909 | 0.7047 | 0.4916 | 0.1654
3 | Linux | PyT | SGD | 1447 | 12.825 | 0.4964 | 68.7747 | 0.7570 | 0.5853 | 0.3531
4 | Linux | PyT | Adam | 1576 | 6.87 | 0.3656 | 114.7003 | 0.6443 | 0.4485 | 0.2172
  |   |   |   |   |   |   |   |   |   |  
5 | Win10 | TF | SGD | 107 | 7.668 | 0.0912 | 814.3304 | 0.2428 | 0.1273 | 0.0064
6 | Win10 | TF | Adam | 53.9 | 0.19 | 0.3647 | 94.6812 | 0.7057 | 0.4980 | 0.1685
7 | Win10 | PyT | SGD | 267423 | 1047 | 0.4978 | 71.0742 | 0.7563 | 0.5839 | 0.3572
8 | Win10 | PyT | Adam | 301242 | 894 | 0.3665 | 114.1199 | 0.6451 | 0.4490 | 0.2181


The takeaways are:
* Windows and Linux are giving effectively the same TransE performance results.  
* For PyTorch, Linux and Windows *loss* values are quite different, whether at Epoch 0 or Epoch 950. Maybe something to explore further.
* I'm now getting very similar results for (Linux, PyT, SGD) that you got, @ShulinCao 

  | OS | FW | Opt | MRR | MR | hit@10 | hit@3 | hit@1
-- | -- | -- | -- | -- | -- | -- | -- | --
Mine | Linux | PyT | SGD | 0.4964 | 68.7747 | 0.7570 | 0.5853 | 0.3531
ShulinCao | Linux | PyT | SGD | 0.49684 | 70.8797 | 0.755895 | 0.584619 | 0.35492

* There clearly is a difference between the TF and PyT versions, whether SGD or Adam (rows 1&3; rows 2&4).  In one case, TF-SGD is (much) worse; in the other case, TF-Adam is about the same, maybe a little better.

Should this be explored further, the difference between OpenKE-Tensorflow and OpenKE-PyTorch, for the same configuration parameters?  Or, is this already a known behavior/issue?

**Comment by AndRossi** [comment]
Hello, I have tried following the example for evaluating TransE on FB15K using TensorFlow.
Some months ago I obtained the same results reported by @ShulinCao.

Very recently, however, I have started getting weird very bad results.

```
no type constraint results:
metric:			 MRR 		 MR 		 hit@10 	 hit@3  	 hit@1 
l(raw):			 0.061542 	 1008.604370 	 0.175518 	 0.070305 	 0.005671 
r(raw):			 0.077907 	 816.915527 	 0.220345 	 0.097290 	 0.005790 
averaged(raw):		 0.069724 	 912.759949 	 0.197931 	 0.083797 	 0.005730 

l(filter):		 0.080521 	 881.159668 	 0.217823 	 0.106380 	 0.006755 
r(filter):		 0.099728 	 736.844971 	 0.270183 	 0.140373 	 0.007398 
averaged(filter):	 0.090125 	 809.002319 	 0.244003 	 0.123377 	 0.007076 
type constraint results:
metric:			 MRR 		 MR 		 hit@10 	 hit@3  	 hit@1 
l(raw):			 0.129656 	 314.202942 	 0.261533 	 0.138867 	 0.062112 
r(raw):			 0.178828 	 228.678726 	 0.334343 	 0.193767 	 0.099761 
averaged(raw):		 0.154242 	 271.440826 	 0.297938 	 0.166317 	 0.080936 

l(filter):		 0.216929 	 186.776825 	 0.377241 	 0.241032 	 0.134618 
r(filter):		 0.254146 	 148.608658 	 0.428095 	 0.281526 	 0.165919 
averaged(filter):	 0.235538 	 167.692749 	 0.402668 	 0.261279 	 0.150268 
The total of test triples is 59071.
```

This is my code (as I said before, it has the same hyperparameters in the test file).

```python
import config
import models
import tensorflow as tf
import numpy as np
import os
os.environ['CUDA_VISIBLE_DEVICES']='0'  # was 7
#Input training files from benchmarks/FB15K/ folder.
con = config.Config()
#True: Input test files from the same folder.
con.set_in_path("./benchmarks/FB15K/")
con.set_test_link_prediction(True)
con.set_test_triple_classification(False)
con.set_work_threads(1) # was 8
con.set_train_times(1000)
con.set_nbatches(100)
con.set_alpha(0.001)
con.set_margin(1.0)
con.set_bern(0)
con.set_dimension(100)
con.set_ent_neg_rate(1)
con.set_rel_neg_rate(0)
con.set_opt_method("SGD")

#Models will be exported via tf.Saver() automatically.
con.set_export_files("./res/model.vec.tf", 0)
#Model parameters will be exported to json files automatically.
con.set_out_files("./res/embedding.vec.json")
#Initialize experimental settings.
con.init()
#Set the knowledge embedding model
con.set_model(models.TransE)
#Train the model.
con.run()
#To test models after training needs "set_test_flag(True)".
con.test()
```

What may be the cause for this?
(This is only partly related to the issue opened by @drcdr so please tell me if you want me to open a separate issue, or we can keep it in this same thread).

**Comment by AndRossi** [comment]
I have done some investigations and I think I have found the cause of my issue.

The problems are caused by the changes introduced with commit [9e51fc2ef60211d1593fc1e0362ed8611a8e6905](https://github.com/thunlp/OpenKE/commit/9e51fc2ef60211d1593fc1e0362ed8611a8e6905) (23 days ago).

I have reverted my TransE.py file to how it was before that commit, and I now obtain good performances again:

```
no type constraint results:
metric:			 MRR 		 MR 		 hit@10 	 hit@3  	 hit@1 
l(raw):			 0.220367 	 269.098846 	 0.464238 	 0.254152 	 0.107565 
r(raw):			 0.277147 	 170.363541 	 0.543651 	 0.318837 	 0.152190 
averaged(raw):		 0.248757 	 219.731201 	 0.503944 	 0.286494 	 0.129878 

l(filter):		 0.472974 	 84.680641 	 0.729478 	 0.558294 	 0.332718 
r(filter):		 0.522775 	 55.392071 	 0.783718 	 0.614786 	 0.378324 
averaged(filter):	 0.497875 	 70.036354 	 0.756598 	 0.586540 	 0.355521 
```

I am not 100% sure of how the changes in that commit can cause such variations in the model performances though. My environment is MacOS, and I'm running TransE on Tensorflow on CPU.
You may want to check these changes in behaviour in your environment as well.

**Comment by drcdr** [comment]
Nice find, @AndRossi,  I think we are both working with the same issue.  

I've been examining the idea of reverting to pre-9e51 a little further.  I've run several test cases today: {SGD, Adam, Adadelta, Adagrad} x {TF-Head, TF-pre-9e51, PyT}.  I've found that OpenKE-PyTorch is usually the same or better that OpenKE-TF.  For TF, Head is worse for SGD and Adadelta, better for Adam, and the same for Adagrad.

I'm only changing these parameters, but otherwise sticking with the scripts above.  So, I don't know how valid these experiments are, since I'm not changing convergence parameters like alpha, train_times, or nbatches.

It seems to me that the PyTorch version is computing the negative_score correctly, averaging over the negative samples in Model.py.  I can't tell if the TF version is doing it correctly or not, in TransE.py, but it seems a little strange to me that the code in loss_def is the same for p_score and n_score, though they are shaped differently.

**Comment by caoh18** [comment]
Hello, I think I also meet  the same problem today. I tried to train and test TranE on FB15k237 but the results are much worse than what previous papers reported. So I tried old version code and the results seemed to be reasonable. I am wondering if there is anything wrong here?

**Comment by MuhibKhan** [comment]
Will you please guide me. I want to check the Head prediction and Tail prediction at (Hits@10) of FB15k dataset. How can I check the different types of relation like 1-to-1, 1-to-N, N-to-1 and N-to-N of FB15k dataset?
Thanks for anticipation

**Comment by THUCSTHanxu13** [comment]
I have provided the scripts for TransE and other models on FB15K237 and WN18RR, which can reach the comparable results as compared with the reported results.

If you want to check the different types of relation like 1-to-1, 1-to-N, N-to-1 and N-to-N of FB15k dataset, you can directly use the file (e.g., 1-1.txt	) instead of test2id.txt.

**Comment by MuhibKhan** [comment]
@THUCSTHanxu13 I want to check the Prediction Head and Prediction Tail (Hits@10) on different types of relation like 1-to-1, 1-to-N, N-to-1 and N-to-N of TransD model like the results mentioned in the paper. How can I perform the Head and Tail Prediction on 1-to-1, 1-to-N, N-to-1 and N-to-N of relations of FB15k dataset?

![TransD](https://user-images.githubusercontent.com/7377986/68836804-2757ac00-06f6-11ea-8151-e75c08a96e90.png)

**Comment by drcdr** [comment]
@MuhibKhan Since your question is off-topic of the original issue, and this issue is closed - suggest opening a new issue.  I can give you a little feedback on the issue there.

**Comment by MuhibKhan** [comment]
Thank you Sir for your response. 
Sorry for my mistake. I create the new issue as you told me in your last mail. 
    On Thursday, November 14, 2019, 10:23:31 PM GMT+8, drcdr <notifications@github.com> wrote:  
 
 
@MuhibKhan Since your question is off-topic of the original issue, and this issue is closed - suggest opening a new issue. I can give you a little feedback on the issue there.
